word-search.py

```diff
@@ -212,9 +216,106 @@ def randomizer(grid):
 
     return grid
 
-def generateWordList():
-    # compiles the list of words that the user can find in the game
-    return
+# creates a dictionary of all distinct letters in the grid as keys
+# and the list of all their coordinate locations as the value
+def find_starting_coordinates(grid):
+    startingCoordinates = {}
+    for x in range(0, len(grid)):
+        for y in range(0, len(grid[x])):
+            if grid[x][y] in startingCoordinates.keys():
+                startingCoordinates[grid[x][y]].append([x, y])
+            else:
+                startingCoordinates[grid[x][y]] = [[x, y]]
+    return startingCoordinates
+
+def find_next_tile(nextLetter, xyCoordinates, grid, usedTiles):
+    x = xyCoordinates[0]
+    y = xyCoordinates[1]
+    maxIndex = len(grid)-1
+    matches = []
+
+    # checks each of the surround tiles if it exists,
+    # if it has not yet been used to form the word,
+    # and if is equal to the next letter in the key/word
+    if (x-1 >= 0 and y-1 >= 0 and (not [x-1, y-1] in usedTiles) and nextLetter == grid[x-1][y-1]):
+        matches.append([x-1, y-1])
+    if (x-1 >= 0 and (not [x-1, y] in usedTiles) and nextLetter == grid[x-1][y]):
+        matches.append([x-1, y])
+    if (x-1 >= 0 and y+1 <= maxIndex and (not [x-1, y+1] in usedTiles) and nextLetter == grid[x-1][y+1]):
+        matches.append([x-1, y+1])
+    if (y-1 >= 0 and (not [x, y-1] in usedTiles) and nextLetter == grid[x][y-1]):
+        matches.append([x, y-1])
+    if (y+1 <= maxIndex and (not [x, y+1] in usedTiles) and nextLetter == grid[x][y+1]):
+        matches.append([x, y+1])
+    if (x+1 <= maxIndex and y-1 >= 0 and (not [x+1, y-1] in usedTiles) and nextLetter == grid[x+1][y-1]):
+        matches.append([x+1, y-1])
+    if (x+1 <= maxIndex and (not [x+1, y] in usedTiles) and nextLetter == grid[x+1][y]):
+        matches.append([x+1, y])
+    if (x+1 <= maxIndex and y+1 <= maxIndex and (not [x+1, y+1] in usedTiles) and nextLetter == grid[x+1][y+1]):
+        matches.append([x+1, y+1])
+
+    # all matching coordinates are noted
+    return matches
+
+def word_mapper(charList, coordinates, grid, usedTiles=[]):
+    if (len(charList) > 1):
+        for c in coordinates:
+            # copy to keep a separate list for each branch of the coordinate paths
+            copy = usedTiles.copy()
+            # append current coordinate used for the letter
+            copy.append(c)
+            # find all tile matched for the next letter
+            tileMatches = find_next_tile(charList[1], c, grid, copy)
+
+            if (len(tileMatches) > 0):
```

Comment:
since both if and else returns something, the other elements in the variable coordinates is not processed. Not sure if ito reason for this bug. 

I tried using this fixed grid:
[['i','t','d','e'],['n','qu','o','c'],['i','e','t','i'],['p','r','a','s']]

supposedly there are 2 branches for the key "qui" 
1st branch should have - quit, quint, quin, etc
2nd branch should have - quiet, quim, etc
and 2 branches for "ni"
1st branch should have - nit
2nd branch should have nip

2nd branch words are not listed in gridWordList

Reply:
Nice catch, and thanks for testing too! :)

Reply:
Yup, we need to continue instead of return False in the else clause inside the for loop. The base case must also return all matches instead of just the first one (line 279). Will update these!

Reply:
Sorry!! Had a few test cases in mind lang that I wanted to test out aside from yung small branch outs ng quit/quiet and nip/nit 😁 Parang last naman na to 😅 

Tried testing a longer word na partially fulfilled by 1 branch pero available sa longer branch.

here are the samples I tried
[['e','i','t','e'],['c','n','qu','c'],['i','i', 'e','t'],['p','r','a','s']]
- nawala na si quiet ulit sa valid words kasi "quie" is partially fulfilled sa upper branch then parang di na naexplore yung lower branch where the word can actually be formed
![image](https://github.com/user-attachments/assets/fad158df-9b5f-4290-9717-c831cdc15b3b)


[['i', 'n', 'y', 'a'], ['t','a','a','a'], ['a','r','t','i'],['a','e','d','n']]
- same here. 2 possible starting points naman for "tin" where "tinder" should be available sa lower right pero di din nahanap 😬  
![image](https://github.com/user-attachments/assets/5b14b571-2bbf-4869-b2d8-d04345c71496)


Hindi ko din 100% gets nangyayari and I know complicated yung part ng project na to 😓 Yung test cases na to yung mga hindi ko gets pano ipprogram eh kaya wanted to test them out. Thank you so much vicky!

Reply:
Thanks for testing thoroughly, Reg! Will note this bug and address this in another branch. In the mean time, let's merge first to prevent conflicts when other PRs come in :)